app/controllers/carts_controller.rb

```diff
@@ -0,0 +1,44 @@
+class CartsController < ApplicationController
+
+  def index
+    @cart = session[:cart]
+  end
+
+  def add
+    product = Product.find(params[:id])
+    product_id = params[:id]
+    quantity = params[:product][:add_quantity].to_i
+    session[:cart] = {} if !session[:cart]
+    cart = session[:cart]
+    if cart[product_id].nil? && product.stock >= quantity # adding product for the first time
```

Comment:
Although minor, I think this `if` clause and the related `elsif` clause could be DRY'd out a little bit:
```
new_quantity = quantity + (cart[product_id] || 0)
if product.stock >= new_quantity
  cart[product_id] = new_quantity
  redirect_to cart_path
else
  flash[:error] = "..."
  redirect_to ...
end
```